🚨 Refactor ViT to updated standards

yonigozlan opened the pull request:
# What does this PR do?

This PR aims at refactoring ViT as part of an effort to standardize vision models in the library, similarly to https://github.com/huggingface/transformers/pull/41549
# Vision Models Refactoring - Chronological Todo List
- [x] `vit` (2021-04-01) #41693
- [x] `deit` (2021-04-13) #41693
- [x] `detr` (2021-06-09) #41549
- [x] `beit` (2021-08-04) #41693
- [x] `segformer` (2021-10-28) #41693
- [ ] `imagegpt` (2021-11-18)
- [x] `vit_mae` (2022-01-18) #41693
- [x] `swin` (2022-01-21) #41693
- [ ] `convnext` (2022-02-07)
- [ ] `poolformer` (2022-02-17)
- [ ] `maskformer` (2022-03-02)
- [ ] `dit` (2022-03-10)
- [ ] `resnet` (2022-03-14)
- [ ] `glpn` (2022-03-22)
- [ ] `dpt` (2022-03-28)
- [ ] `regnet` (2022-04-07)
- [ ] `yolos` (2022-05-02)
- [ ] `cvt` (2022-05-18)
- [ ] `levit` (2022-06-01)
- [ ] `mobilevit` (2022-06-29)
- [ ] `swinv2` (2022-07-27)
- [x] `deformable_detr` (2022-09-14) #41549
- [x] `conditional_detr` (2022-09-22) #41549
- [x] `vit_msn` (2022-09-22) #41693
- [ ] `table-transformer` (2022-10-18)
- [ ] `mobilenet_v2` (2022-11-14)
- [ ] `dinat` (2022-11-18)
- [ ] `mobilenet_v1` (2022-11-21)
- [ ] `bit` (2022-12-07)
- [ ] `swin2sr` (2022-12-16)
- [ ] `mask2former` (2023-01-16)
- [ ] `upernet` (2023-01-16)
- [ ] `efficientnet` (2023-02-20)
- [ ] `convnextv2` (2023-03-14)
- [ ] `sam` (2023-04-19)
- [ ] `focalnet` (2023-04-23)
- [ ] `swiftformer` (2023-05-12)
- [ ] `mobilevitv2` (2023-06-02)
- [ ] `dinov2` (2023-07-18)
- [ ] `pvt` (2023-07-24)
- [ ] `vitdet` (2023-08-29)
- [ ] `vitmatte` (2023-09-19)
- [ ] `depth_anything` (2024-01-25)
- [ ] `seggpt` (2024-02-26)
- [ ] `pvt_v2` (2024-03-13)
- [ ] `superpoint` (2024-03-19)
- [x] `rt_detr` (2024-06-22) #41549
- [ ] `depth_anything_v2` (2024-07-05)
- [ ] `zoedepth` (2024-07-08)
- [ ] `hiera` (2024-07-12)
- [x] `ijepa` (2024-12-05) #41693
- [ ] `dinov2_with_registers` (2024-12-24)
- [ ] `textnet` (2025-01-08)
- [ ] `vitpose` (2025-01-08)
- [ ] `superglue` (2025-01-20)
- [ ] `dab-detr` (2025-02-04)
- [x] `rt_detr_v2` (2025-02-06) #41549
- [ ] `depth_pro` (2025-02-10)
- [ ] `prompt_depth_anything` (2025-03-21)
- [ ] `mlcd` (2025-04-15)
- [ ] `sam_hq` (2025-04-28)
- [x] `d_fine` (2025-04-29) #41549
- [ ] `hgnet_v2` (2025-04-29)
- [ ] `lightglue` (2025-06-17)
- [ ] `eomt` (2025-06-27)
- [ ] `aimv2` (2025-07-08)
- [ ] `efficientloftr` (2025-07-22)
- [ ] `dinov3` (2025-08-14)
- [ ] `sam2` (2025-08-14)
- [ ] `sam3_tracker` (2025-11-19)
- [x] `pixio` (2025-12-16) #41693
- [ ] `lw_detr` (2026-01-12)
- [ ] `eomt_dinov3` (2026-02-01)
- [ ] `chmv2` (2026-03-11)

**Total: 74 vision models**

Additional multimodal/other modalities refactor:
- [x] `audio_spectrogram_transformer` #41693
- [x] `vivit` #41693
- [x] `pp_doclayout_v3` #41549

yonigozlan committed:
refactor attention

yonigozlan committed:
use mlp instead of intermediate and output

yonigozlan commented:
Cc @vasqu for attention mask!

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_41693). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

vasqu commented on src/transformers/models/vit/modeling_vit.py:
```diff
@@ -362,7 +330,15 @@ class ViTPreTrainedModel(PreTrainedModel):
+        "attention.output.dense": "attention.o_proj",
+        "intermediate.dense": "mlp.fc1",
+        "output.dense": "mlp.fc2",
     }
```
Ah, I see so we rename stuff on the fly. I'm unsure about this would like an opinion from @zucchini-nlp or @molbap here

vasqu reviewed (commented):
Looking already pretty good to me, I'm just really unsure if we want to have the conversion mapping; would wait for the others on their opinion here

yonigozlan committed:
use modular for deit

yonigozlan commented:
Thanks for the review!
> Looking already pretty good to me, I'm just really unsure if we want to have the conversion mapping; would wait for the others on their opinion here

It seems reasonable for v5 imo, and will be well worth the cost in the long run. + @ArthurZucker and @Cyrilvallez are working on a way to make weights conversion on the fly much faster

yonigozlan committed:
nit

yonigozlan committed:
fix deit (to check) and refactor beit

yonigozlan committed:
use modular for beit

molbap reviewed (commented):
daang I had published my review yesterday but it disappeared 😩

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into refactor-vit

yonigozlan committed:
style

yonigozlan committed:
refactor vit deit beit segformer

yonigozlan committed:
add pos_embed_utils, refactor swin

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into refactor-vit

yonigozlan committed:
Fix after merge + refactor pixio

yonigozlan committed:
refactor most drop_path in transformers

yonigozlan committed:
refactor vit_msn

yonigozlan committed:
fix modular

yonigozlan committed:
refactor vivit and audio_spectrogram_transformer, remove obsolete copied from and add todos

yonigozlan committed:
Fix repo

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into refactor-vit

yonigozlan committed:
fix nit

yonigozlan committed:
Fix drop_path_schedule

yonigozlan committed:
nit

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into refactor-vit

Cyrilvallez commented on src/transformers/core_model_loading.py:
```diff
@@ -113,15 +113,24 @@ def convert(
+            # Extract layer index from full_layer_name (e.g. "transformer.h.5.self_attn.q_proj.weight" -> "5")
+            match = re.search(r"\.(\d+)\.", full_layer_name)
+            if match:
+                layer_idx = match.group(1)
```
What are all those changes in `Chunk`? On the new conversions, I see only weight renamings, so there should not be any motivation for such a change 🤔

Cyrilvallez reviewed (commented):
Very nice! Did not check all models in details for now, but those I checked look good! For now I mostly focused on more general aspects. Especially, I think we break too much from our philosophy of self-contained modeling files here - let's keep general functions/modules live in the modelings instead of being general (same as with do with `eager_attention_forward`), and simply import them with modular!

Also, since we are changing a lot of weight names, please let's run `run-slow` on ALL models that have a new weight mapping -> a single small typo can fully break any model. Would be even nicer to run on all touched models to be sure as some have very big architectural changes!

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into refactor-vit

yonigozlan committed:
use modular for pos embed

yonigozlan committed:
reverse DropPath as a util

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into refactor-vit

yonigozlan committed:
add conversion ijepa + fix offload vit_mae

yonigozlan committed:
fix modular

yonigozlan committed:
fix qkv bias in output proj + swin attention per stage

vasqu commented on src/transformers/models/beit/modular_beit.py:
````diff
@@ -0,0 +1,1071 @@
+        ```"""
+        if labels is not None and self.config.num_labels == 1:
+            raise ValueError("The number of labels should be greater than one")
+        output_hidden_states = kwargs.pop("output_hidden_states", self.config.output_hidden_states)
````
Check out backbone utils, we have filter output hidden states decorator - and we force this within kwargs then

vasqu commented on src/transformers/models/swin/modular_swin.py:
```diff
@@ -0,0 +1,1122 @@
+        # [batch_size, height/2, width/2, num_channels]
+        input_feature_2 = input_feature[:, 0::2, 1::2, :]
+        # [batch_size, height/2, width/2, num_channels]
+        input_feature_3 = input_feature[:, 1::2, 1::2, :]
```
Could be a loop instead over the indices maybe? It's essentially rotating

vasqu reviewed (commented):
I went through the more important models first to give some important feedback. I'm not sure what it was all about with each own file etc (I guess this is no longer the case)

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into refactor-vit

yonigozlan committed:
address review comments

yonigozlan committed:
fix swinn

yonigozlan committed:
improve swin and beit

yonigozlan commented on src/transformers/models/vit/modeling_vit.py:
```diff
@@ -356,14 +301,30 @@ def __init__(self, config: ViTConfig):
         return BaseModelOutput(last_hidden_state=hidden_states)
 
 
+class ViTPooler(nn.Module):
```
Same as above about not using modular for vit

yonigozlan commented on src/transformers/core_model_loading.py:
```diff
@@ -116,7 +116,8 @@ def convert(
         chunks = torch.chunk(tensor, sizes, dim=self.dim)
-        return dict(zip(targets, chunks))
+        # Clone to ensure contiguous tensors (chunk returns views that may be non-contiguous)
+        return dict(zip(targets, [c.contiguous().clone() for c in chunks]))
```
That's a claude improvisation that I didn't catch. Not sure tbh, I'll remove it as it's out of scope for this PR, but it might be a true issue.

yonigozlan commented on src/transformers/models/swin/modular_swin.py:
```diff
@@ -0,0 +1,1122 @@
+    </Tip>
+    """
+)
+class SwinForImageClassification(SwinPreTrainedModel):
```
Not sure yet but yes I can see this becoming a pattern that we can inherit more and more as we keep refactoring these vision models. I don't think there's a way to inherit these cleanly for now though

yonigozlan committed:
last fixes after review

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into refactor-vit

yonigozlan committed:
standardize backbone beit

yonigozlan committed:
fix beit

yonigozlan committed:
fix beit

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into refactor-vit

yonigozlan committed:
fix repo

vasqu commented on src/transformers/models/vit/modeling_vit.py:
```diff
@@ -196,172 +192,114 @@ def eager_attention_forward(
-    layernorm applied before each block.
-    """
-
+class ViTMLP(nn.Module):
```
Just so I understood it correctly. ViT stays completely clean of any inheritance and we have it act as golden standard - correct?

yonigozlan replied:
yes that's the idea!

vasqu commented on src/transformers/models/vit/modeling_vit.py:
```diff
@@ -196,172 +192,114 @@ def eager_attention_forward(
-        # in ViT, layernorm is also applied after self-attention
-        layer_output = self.layernorm_after(hidden_states)
-        layer_output = self.intermediate(layer_output)
+        residual = hidden_states
```
Super nit: can we also have the same comments like # Self Attention and # Fully Connected like in llama

vasqu commented on src/transformers/models/beit/modular_beit.py:
```diff
@@ -0,0 +1,1021 @@
+            nn.Parameter(init_values * torch.ones(config.hidden_size), requires_grad=True) if init_values > 0 else 1.0
+        )
+        self.relative_position_bias = BeitRelativePositionBias(config) if config.use_relative_position_bias else None
+        self.dropout = nn.Dropout(config.hidden_dropout_prob)
```
Can we not inherit from ViT, at least for the init or will this become too messy?

vasqu commented on src/transformers/models/beit/modular_beit.py:
```diff
@@ -0,0 +1,1021 @@
+
+
+@auto_docstring
+class BeitPreTrainedModel(PreTrainedModel):
```
Do we want to avoid inheritance? Imo, we can definitely use vit an overwrite / add onto it. It will avoid this flag mess

vasqu commented on src/transformers/models/swin/modular_swin.py:
```diff
@@ -0,0 +1,1115 @@
+
+class SwinEmbeddings(nn.Module):
+    """
+    Construct the patch and position embeddings. Optionally, also the mask token.
```
Not similar to anything?

vasqu replied:
Thanks, makes sense then. It is hard to keep track of this from an outside perspective but totally fair to establish a proper order from whom to inherit from

vasqu reviewed (commented):
I went through the 3 core models again. Looks already much better imo, lmk if there are other special ones. I think after this round, I would make a bigger round but then we should be ready ish 🫡

yonigozlan committed:
changes after review

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into refactor-vit

yonigozlan committed:
Fix beit

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into refactor-vit

vasqu commented on src/transformers/loss/loss_utils.py:
```diff
@@ -142,7 +142,27 @@ def ForTokenClassification(logits: torch.Tensor, labels, config, **kwargs):
+    **kwargs,
+) -> torch.Tensor:
+    upsampled_logits = nn.functional.interpolate(logits, size=labels.shape[-2:], mode="bilinear", align_corners=False)
+    loss_fct = CrossEntropyLoss(ignore_index=ignore_index)
```
We should use `fixed_cross_entropy` instead and properly allow num items in batch

vasqu commented on src/transformers/models/vit/modeling_vit.py:
```diff
@@ -430,49 +369,36 @@ def forward(
         pixel_values: torch.Tensor | None = None,
         bool_masked_pos: torch.BoolTensor | None = None,
         interpolate_pos_encoding: bool | None = None,
+        attention_mask: torch.Tensor | None = None,
```
Gotcha, just a bit wondering what future models should do: attention mask or pixel attention mask.

vasqu commented on src/transformers/models/beit/modular_beit.py:
```diff
@@ -0,0 +1,987 @@
+
+
+@auto_docstring
+class BeitPreTrainedModel(ViTPreTrainedModel):
```
I think we should be careful about FA and Flex attention support, especially because of the relative biases

vasqu commented on src/transformers/models/beit/modular_beit.py:
```diff
@@ -0,0 +1,1021 @@
+            dilation=dilation,
+        )
+        self.bn = nn.BatchNorm2d(out_channels)
+        self.activation = nn.ReLU()
```
I mean we can override the init and keep the forward, no?

yonigozlan replied:
Tried to standardize this a bit, should be better now! (inheriting from resnet)

vasqu commented on src/transformers/models/beit/modular_beit.py:
```diff
@@ -0,0 +1,987 @@
+            upsampled = nn.functional.interpolate(
+                pooled, size=hidden_states.size()[2:], mode="bilinear", align_corners=False
+            )
+            outputs.append(upsampled)
```
Imo, we can move the interpolaton inside the block, no? I don't see the module used anywhere else tbh

vasqu commented on src/transformers/models/beit/modular_beit.py:
```diff
@@ -0,0 +1,987 @@
+            self.in_channels[-1],
+            self.channels,
+        )
+        self.bottleneck = BeitConvModule(
```
should we rename with psp prefix for that bottleneck?

vasqu commented on src/transformers/models/beit/modular_beit.py:
```diff
@@ -0,0 +1,987 @@
+        )
+
+    def psp_forward(self, hidden_states: list[torch.Tensor]) -> torch.Tensor:
+        x = hidden_states[-1]
```
```suggestion
        hidden_state = hidden_states[-1]
```

vasqu commented on src/transformers/models/beit/modular_beit.py:
```diff
@@ -0,0 +1,987 @@
+        # build laterals
+        laterals = [
+            lateral_conv(hidden_state) for lateral_conv, hidden_state in zip(self.lateral_convs, encoder_hidden_states)
+        ]
```
Lets make this a normal loop please instead of a list comprehension

vasqu commented on src/transformers/models/beit/modular_beit.py:
```diff
@@ -0,0 +1,987 @@
+        if self.concat_input:
+            output = self.conv_cat(torch.cat([hidden_states, output], dim=1))
+        output = self.classifier(output)
+        return output
```
```suggestion
        residual = encoder_hidden_states[self.in_index]
        hidden_states = residual
        for conv in self.convs:
            hidden_states = conv(hidden_states)
        if self.concat_input:
            hidden_states = self.conv_cat(torch.cat([residual, hidden_states], dim=1))
        hidden_states = self.classifier(hidden_states)
        return hidden_states
```
nit: but would just use hidden states and residual whenever possible, no need for "new" namings

vasqu commented on src/transformers/models/beit/modular_beit.py:
```diff
@@ -0,0 +1,987 @@
+        self.conv_transpose1 = nn.ConvTranspose2d(hidden_size, hidden_size, kernel_size=2, stride=2)
+        self.bn = nn.BatchNorm2d(hidden_size)
+        self.activation = nn.GELU()
+        self.conv_transpose2 = nn.ConvTranspose2d(hidden_size, hidden_size, kernel_size=2, stride=2)
```
Would be nice if we could make this more configurable for future models --> kernel size, stride as args for the init

vasqu commented on src/transformers/models/beit/modular_beit.py:
```diff
@@ -0,0 +1,987 @@
+
+    def __init__(self, config: BeitConfig):
+        super().__init__()
+        hidden_size = config.hidden_size
```
Imo, we can remove that and just use the config directly?

vasqu commented on src/transformers/models/beit/modular_beit.py:
```diff
@@ -0,0 +1,987 @@
+        hidden_size = config.hidden_size
+        self.fpn1 = BeitFPNUpBlock(hidden_size)
+        self.fpn2 = nn.ConvTranspose2d(hidden_size, hidden_size, kernel_size=2, stride=2)
+        self.fpn4 = nn.MaxPool2d(kernel_size=2, stride=2)
```
```suggestion
        self.fpn3 = nn.MaxPool2d(kernel_size=2, stride=2)
```
a bit confusing otherwise imo

yonigozlan replied:
the 4 is because this is applied to the 4th input feature maps, so this still makes sense imo. fpn3 was removed here because it was identity.

vasqu commented on src/transformers/models/beit/modular_beit.py:
```diff
@@ -0,0 +1,987 @@
+        # out_indices are 1-based into encoder_hidden_states (index 0 is the initial patch embedding).
+        # Remove the CLS token ([:, 1:]) and reshape from sequence to 2D spatial feature maps.
+        feature_maps = tuple(
+            encoder_hidden_states[i - 1][:, 1:].permute(0, 2, 1).reshape(batch_size, -1, patch_height, patch_width)
```
```suggestion
            encoder_hidden_states[i - 1][:, 1:].transpose(1, 2).reshape(batch_size, -1, patch_height, patch_width)
```

vasqu commented on src/transformers/models/beit/modular_beit.py:
```diff
@@ -0,0 +1,987 @@
+        self.encoder = BeitModel(config, add_pooling_layer=False)
+
+        if config.add_fpn:
+            self.fpn = BeitFPNNeck(config)
```
ternary with identity might be cleaner imo

vasqu commented on src/transformers/models/beit/modular_beit.py:
```diff
@@ -0,0 +1,987 @@
+        super().__init__(config)
+
+        self.num_features = [config.hidden_size for _ in range(config.num_hidden_layers + 1)]
+        self.encoder = BeitModel(config, add_pooling_layer=False)
```
Should be the base model prefix instead of encoder no?

vasqu commented on src/transformers/models/beit/modular_beit.py:
```diff
@@ -0,0 +1,987 @@
+        self.post_init()
+
+    def get_input_embeddings(self):
+        return self.encoder.embeddings.patch_embeddings
```
Should be possible via `self.encoder.get_input_embeddings()` no? Maybe we dont even need to override this anymore at all but unsure

vasqu commented on src/transformers/models/swin/modular_swin.py:
```diff
@@ -0,0 +1,1106 @@
+        self.image_size = image_size
+        self.patch_size = patch_size
+        self.num_channels = num_channels
+        self.num_patches = num_patches
```
Might be wrong, but I dont think we need all the attributes as class attributes

vasqu commented on src/transformers/models/swin/modular_swin.py:
```diff
@@ -0,0 +1,1106 @@
+
+        return relative_coords.sum(-1)  # Wh*Ww, Wh*Ww
+
+    def forward(self) -> torch.Tensor:
```
Hmm, this is a static result. Wouldn't it make sense to buffer this on init within attention so you dont need to call the forward all the time

yonigozlan replied:
But `self.relative_position_bias_table` is learnable, so this call changes at every pass during training

vasqu commented on src/transformers/models/swin/modular_swin.py:
```diff
@@ -0,0 +1,1106 @@
+        self.window_size = config.window_size
+        super().__init__()
+        self.attention = SwinAttention(config, dim, num_heads, window_size=self.window_size)
+        self.layernorm_before = nn.LayerNorm(dim, eps=config.layer_norm_eps)
+        self.layernorm_after = nn.LayerNorm(dim, eps=config.layer_norm_eps)
```
```suggestion
```
should be inherited no?

yonigozlan replied:
We can't because we have dim (which changes) here instead of config.hidden_size

vasqu commented on src/transformers/models/swin/modular_swin.py:
```diff
@@ -0,0 +1,1115 @@
+        hidden_states = self.mlp(hidden_states)
+        hidden_states = self.dropout(hidden_states) + residual
+
+        return hidden_states, attn_weights
```
Ah sorry, I missed that connection

vasqu commented on src/transformers/models/swin/modular_swin.py:
```diff
@@ -0,0 +1,1106 @@
+class SwinPreTrainedModel(ViTPreTrainedModel):
+    config: SwinConfig
+    _no_split_modules = ["SwinStage"]
+    _supports_flash_attn = False
```
Also flex attn I think to disable

vasqu commented on src/transformers/models/swin/modular_swin.py:
```diff
@@ -0,0 +1,1106 @@
+                    drop_path=dpr[sum(config.depths[:i_layer]) : sum(config.depths[: i_layer + 1])],
+                    downsample=SwinPatchMerging if (i_layer < self.num_layers - 1) else None,
+                )
+                for i_layer in range(self.num_layers)
```
can we have a different name for i layer, kind of sounds off to me

vasqu commented on src/transformers/models/swin/modular_swin.py:
```diff
@@ -0,0 +1,1106 @@
+
+        self.num_features = [config.embed_dim] + [int(config.embed_dim * 2**i) for i in range(len(config.depths))]
+        self.embeddings = SwinEmbeddings(config)
+        self.encoder = SwinEncoder(config, self.embeddings.patch_grid)
```
Same here as in beit, should be probably the base model prefix instead of encoder

vasqu reviewed (commented):
Ok, so I just wanted to go through the "big 3". I will check the other models (in a less detailed way) next week then so we can get closer to merging 🤗 sorry for the delays and all

yonigozlan commented on src/transformers/models/vit/modeling_vit.py:
```diff
@@ -196,119 +193,72 @@ def eager_attention_forward(
-        self.all_head_size = self.num_attention_heads * self.attention_head_size
-        self.dropout_prob = config.attention_probs_dropout_prob
-        self.scaling = self.attention_head_size**-0.5
+        self.head_dim = config.hidden_size // config.num_attention_heads
```
added!

yonigozlan committed:
fix after review

yonigozlan commented:
Thanks again for the review @vasqu ! Addressed your comments ;)

yonigozlan committed:
fix unexpected key swin

vasqu commented on src/transformers/loss/loss_utils.py:
```diff
@@ -142,7 +142,31 @@ def ForTokenClassification(logits: torch.Tensor, labels, config, **kwargs):
+        upsampled_auxiliary_logits = nn.functional.interpolate(
+            auxiliary_logits, size=labels.shape[-2:], mode="bilinear", align_corners=False
+        )
+        loss += auxiliary_loss_weight * fixed_cross_entropy(
```
Nit: would make it out of place, inplace mods are always a bit more risky on torch side

vasqu commented on src/transformers/models/audio_spectrogram_transformer/modular_audio_spectrogram_transformer.py:
```diff
@@ -0,0 +1,264 @@
+    @torch.no_grad()
+    def _init_weights(self, module):
+        """Initialize the weights"""
+        if isinstance(module, (nn.Linear, nn.Conv2d)):
+            init.trunc_normal_(module.weight, mean=0.0, std=self.config.initializer_range)
+            if module.bias is not None:
+                init.zeros_(module.bias)
+        elif isinstance(module, nn.LayerNorm):
+            init.zeros_(module.bias)
+            init.ones_(module.weight)
```
We can use super for these no? Meaning the base PreTrained model one not vit's

vasqu commented on src/transformers/models/audio_spectrogram_transformer/modular_audio_spectrogram_transformer.py:
```diff
@@ -0,0 +1,264 @@
+    output) e.g. for datasets like AudioSet, Speech Commands v2.
+    """
+)
+class ASTForAudioClassification(ASTPreTrainedModel):
```
Not for here but could also be moved to a generic layer class like we did for causal lms

vasqu commented on src/transformers/models/d_fine/modeling_d_fine.py:
```diff
@@ -610,6 +610,54 @@ def forward(
         return hidden_states
 
 
+def build_2d_sinusoidal_position_embedding(
```
Just a general question, when do we have these functions as general ones and when within the module itself? There doesn't seem to be too much concistency - don't think it's that big of a deal but still interested on your thoughts

yonigozlan replied:
Yes there's probably some incoherences in some modules to be fixed, although the ones that were refactored should all have build_2d_sinusoidal_position_embedding as general functions

vasqu commented on src/transformers/models/dab_detr/modeling_dab_detr.py:
```diff
@@ -37,6 +37,42 @@
 logger = logging.get_logger(__name__)
 
 
+def encode_sinusoidal_position_embedding(
```
Same as conditional detr? Maybe a temporary copied from statement to keep track

vasqu commented on src/transformers/models/data2vec/modeling_data2vec_vision.py:
```diff
@@ -56,38 +56,7 @@ class Data2VecVisionModelOutputWithPooling(BaseModelOutputWithPooling):
-
-
-# Copied from transformers.models.beit.modeling_beit.BeitEmbeddings with Beit->Data2VecVision
+# Todo - Refactor as part of vision refactor. Copied from transformers.models.beit.modeling_beit.BeitEmbeddings with Beit->Data2VecVision
```
Fmi, so this will be part of a future refactor batch? Just because it's mostly beit

yonigozlan replied:
Yes, just want to get the current ones merged as the PR is huge already

vasqu commented on src/transformers/models/deit/modular_deit.py:
```diff
@@ -0,0 +1,306 @@
+from dataclasses import dataclass
```
licence missing

vasqu commented on src/transformers/models/deit/modular_deit.py:
```diff
@@ -0,0 +1,306 @@
+from .configuration_deit import DeiTConfig
+
+
+class DeiTEmbeddings(ViTEmbeddings):
```
We inherit but override everything it seems; could we highlight those diffs with small comment(s)

vasqu commented on src/transformers/models/dinov2/modeling_dinov2.py:
```diff
@@ -178,7 +178,7 @@ def eager_attention_forward(
 
 
-# Copied from transformers.models.vit.modeling_vit.ViTSelfAttention with ViT->Dinov2
+# Todo - Refactor as part of vision refactor. Copied from transformers.models.vit.modeling_vit.ViTAttention with ViT->Dinov2
```
Ok so I guess these connections are too deep / the PR is big enough

yonigozlan replied:
Yes imo, otherwise we'll never get anything merged 😅

vasqu commented on src/transformers/models/grounding_dino/modeling_grounding_dino.py:
```diff
@@ -35,6 +35,42 @@
 logger = logging.get_logger(__name__)
 
 
+def encode_sinusoidal_position_embedding(
```
Same re copied from might be a good choice to keep track

vasqu commented on src/transformers/models/lw_detr/modular_lw_detr.py:
```diff
@@ -97,6 +97,9 @@ class LwDetrViTConfig(VitDetConfig):
 
     def __post_init__(self, **kwargs):
         self.num_windows_side = int(math.sqrt(self.num_windows))
+        # ViTAttention expects these; map from dropout_prob
+        self.attention_probs_dropout_prob = self.dropout_prob
+        self.hidden_dropout_prob = self.dropout_prob
```
Wouldn't it make more sense to override the init in those cases, this is a weird remapping and might be more confusing tbh.

vasqu commented on src/transformers/models/pixio/modular_pixio.py:
```diff
@@ -261,8 +237,28 @@ def forward(
+
+        self.post_init()
+
+    def get_input_embeddings(self) -> PixioPatchEmbeddings:
+        return self.encoder.embeddings.patch_embeddings
```
Should we rather call with `self.encoder.get_input_embeddings()`?

vasqu commented on src/transformers/models/pixio/modular_pixio.py:
```diff
@@ -261,8 +237,28 @@ def forward(
+        super().__init__(config)
+
+        self.num_features = [config.hidden_size for _ in range(config.num_hidden_layers + 1)]
+        self.encoder = PixioModel(config)
```
Feel like I might have talked about it but I guess we don't use base model prefix here? Or just use it as standard to always use encoder?

yonigozlan replied:
We should honestly, I'll change the weight name

vasqu commented on src/transformers/models/segformer/modular_segformer.py:
```diff
@@ -247,4 +260,556 @@ def _preprocess(
+    attentions: tuple[torch.FloatTensor] | None = None
+
+
+class SegformerOverlapPatchEmbeddings(nn.Module):
```
Fyi, layoutlmv3 looks similar in nature just slightly different passing of args / configs

yonigozlan replied:
Noted for future refactoring :)

vasqu commented on src/transformers/models/segformer/modular_segformer.py:
```diff
@@ -247,4 +260,556 @@ def _preprocess(
+        self.config = config
+        self.num_attention_heads = num_attention_heads
+
+        if hidden_size % num_attention_heads != 0:
+            raise ValueError(
+                f"The hidden size ({hidden_size}) is not a multiple of the number of attention "
+                f"heads ({num_attention_heads})"
+            )
```
```suggestion
```
should really not be needed, if anything a check in the config

vasqu commented on src/transformers/models/segformer/modular_segformer.py:
```diff
@@ -247,4 +260,556 @@ def _preprocess(
+        self.v_proj = nn.Linear(hidden_size, num_attention_heads * self.head_dim)
+        self.o_proj = nn.Linear(num_attention_heads * self.head_dim, hidden_size)
+
+        self.sr_ratio = sequence_reduction_ratio
+        if sequence_reduction_ratio > 1:
+            self.sr = nn.Conv2d(
+                hidden_size, hidden_size, kernel_size=sequence_reduction_ratio, stride=sequence_reduction_ratio
+            )
+            self.layer_norm = nn.LayerNorm(hidden_size)
```
The init looks fairly standard to me, maybe we can use modular just for the init sake and only introduce this here after the super

vasqu commented on src/transformers/models/segformer/modular_segformer.py:
```diff
@@ -247,4 +260,556 @@ def _preprocess(
+        # Sequence reduction: reshape to spatial, apply strided conv, reshape back
+        kv_hidden_states = hidden_states
+        if self.sr_ratio > 1:
+            batch_size, seq_len, num_channels = hidden_states.shape
+            kv_hidden_states = hidden_states.permute(0, 2, 1).reshape(batch_size, num_channels, height, width)
+            kv_hidden_states = self.sr(kv_hidden_states)
+            kv_hidden_states = kv_hidden_states.reshape(batch_size, num_channels, -1).permute(0, 2, 1)
```
This is a bit awkward, could we make a module around the conv2d and add these special reshapes there + let's use transpose ;)

vasqu commented on src/transformers/models/segformer/modular_segformer.py:
```diff
@@ -247,4 +260,556 @@ def _preprocess(
+            kv_hidden_states = kv_hidden_states.reshape(batch_size, num_channels, -1).permute(0, 2, 1)
+            kv_hidden_states = self.layer_norm(kv_hidden_states)
+
+        kv_input_shape = kv_hidden_states.shape[:-1]
+        kv_hidden_shape = (*kv_input_shape, -1, self.head_dim)
```
```suggestion
        kv_hidden_shape = (*kv_hidden_states.shape[:-1], -1, self.head_dim)
```
we dont use the input shape of kv

vasqu commented on src/transformers/models/segformer/modular_segformer.py:
```diff
@@ -247,4 +260,556 @@ def _preprocess(
+
+        attn_output = attn_output.reshape(*input_shape, -1).contiguous()
+        attn_output = self.o_proj(attn_output)
+        attn_output = nn.functional.dropout(attn_output, p=self.hidden_dropout, training=self.training)
```
Would move the dropout outside to the parent tbh

vasqu commented on src/transformers/models/segformer/modular_segformer.py:
```diff
@@ -247,4 +260,556 @@ def _preprocess(
+        return attn_output, attn_weights
+
+
+class SegformerDWConv(nn.Module):
```
Wait isn't that perfect for the sequence_reduction module within the attention 👀

yonigozlan replied:
Close but not exactly (kernel/stride, groups etc.)

vasqu commented on src/transformers/models/segformer/modular_segformer.py:
```diff
@@ -247,4 +260,556 @@ def _preprocess(
+        super().__init__()
+        out_features = out_features or in_features
+        self.fc1 = nn.Linear(in_features, hidden_features)
+        self.dwconv = SegformerDWConv(hidden_features)
```
Can we explicit about the DW with DepthWise?

vasqu commented on src/transformers/models/segformer/modular_segformer.py:
```diff
@@ -247,4 +260,556 @@ def _preprocess(
+        return hidden_states
+
+
+class SegformerMixMLP(nn.Module):
```
Noticing `PvtV2` is closely related to the model as a heads up for future possible refactors

vasqu commented on src/transformers/models/segformer/modular_segformer.py:
```diff
@@ -247,4 +260,556 @@ def _preprocess(
+        )
+        self.drop_path = SegformerDropPath(drop_path) if drop_path > 0.0 else nn.Identity()
+        self.layernorm_after = nn.LayerNorm(hidden_size)
+        mlp_hidden_size = int(hidden_size * mlp_ratio)
+        self.mlp = SegformerMixMLP(config, in_features=hidden_size, hidden_features=mlp_hidden_size)
```
```suggestion
        self.mlp = SegformerMixMLP(config, in_features=hidden_size, hidden_features=int(hidden_size * mlp_ratio))
```
nit

vasqu commented on src/transformers/models/segformer/modular_segformer.py:
```diff
@@ -247,4 +260,556 @@ def _preprocess(
+        width: int,
+        **kwargs: Unpack[TransformersKwargs],
+    ) -> torch.Tensor:
+        residual = hidden_states
+        hidden_states, _ = self.attention(self.layernorm_before(hidden_states), height, width, **kwargs)
+        hidden_states = self.drop_path(hidden_states) + residual
+
+        residual = hidden_states
+        hidden_states = self.mlp(self.layernorm_after(hidden_states), height, width)
+        hidden_states = self.drop_path(hidden_states) + residual
```
```suggestion
        residual = hidden_states
        hidden_states = self.layernorm_before(hidden_states)
        hidden_states, _ = self.attention(hidden_states, height, width, **kwargs)
        hidden_states = self.drop_path(hidden_states) + residual

        residual = hidden_states
        hidden_states = self.layernorm_after(hidden_states)
        hidden_states = self.mlp(hidden_states, height, width)
        hidden_states = self.drop_path(hidden_states) + residual
```
nit: just a personal preference to avoid nesting calls

vasqu commented on src/transformers/models/segformer/modular_segformer.py:
```diff
@@ -247,4 +260,556 @@ def _preprocess(
+        )
+        self.blocks = nn.ModuleList(
+            [
+                SegformerLayer(
+                    config,
+                    hidden_size=config.hidden_sizes[stage_idx],
+                    num_attention_heads=config.num_attention_heads[stage_idx],
+                    drop_path=drop_path_decays[depth_start + layer_idx],
+                    sequence_reduction_ratio=config.sr_ratios[stage_idx],
+                    mlp_ratio=config.mlp_ratios[stage_idx],
+                )
+                for layer_idx in range(config.depths[stage_idx])
```
Ig this was this way already but has each stage really its own config for all these 👀

yonigozlan replied:
Yes looks like it!

vasqu commented on src/transformers/models/segformer/modular_segformer.py:
```diff
@@ -247,4 +260,556 @@ def _preprocess(
+        hidden_states: torch.Tensor,
+        **kwargs: Unpack[TransformersKwargs],
+    ) -> torch.Tensor:
+        hidden_states, height, width = self.patch_embeddings(hidden_states)
+        for block in self.blocks:
+            hidden_states = block(hidden_states, height, width, **kwargs)
+        hidden_states = self.layer_norm(hidden_states)
+        if self.reshape:
+            batch_size = hidden_states.shape[0]
+            hidden_states = hidden_states.reshape(batch_size, height, width, -1).permute(0, 3, 1, 2).contiguous()
```
spacing

vasqu commented on src/transformers/models/segformer/modular_segformer.py:
```diff
@@ -247,4 +260,556 @@ def _preprocess(
+        return hidden_states
+
+
+class SegformerEncoder(nn.Module):
```
This module seems useless to me: It is essentially the same encoder structure we have often seen where we only have a module list we iterate through. Can we move this to the parent module instead

vasqu commented on src/transformers/models/segformer/modular_segformer.py:
```diff
@@ -247,4 +260,556 @@ def _preprocess(
+    main_input_name = "pixel_values"
+    input_modalities = ("image",)
+    supports_gradient_checkpointing = True
+    _no_split_modules = None
```
```suggestion
    _no_split_modules = []
```
Tbh there should be the layer and stages but unsure (would need testing)

vasqu commented on src/transformers/models/segformer/modular_segformer.py:
```diff
@@ -247,4 +260,556 @@ def _preprocess(
+
+
+@auto_docstring
+class SegformerPreTrainedModel(PreTrainedModel):
```
Not important would appreciate some modular inheritance tbh - it's hard to keep track of all flags otherwise if we refactor further

vasqu commented on src/transformers/models/segformer/modular_segformer.py:
```diff
@@ -247,4 +260,556 @@ def _preprocess(
+        super().__init__(config)
+        self.config = config
+
+        # hierarchical Transformer encoder
+        self.encoder = SegformerEncoder(config)
```
Yea exactly, this is super useless

vasqu commented on src/transformers/models/segformer/modular_segformer.py:
```diff
@@ -247,4 +260,556 @@ def _preprocess(
+        # linear layers which will unify the channel dimension of each of the encoder blocks to the same config.decoder_hidden_size
+        linear_projections = []
+        for stage_idx in range(config.num_encoder_blocks):
+            linear_proj = SegformerMLP(config, input_dim=config.hidden_sizes[stage_idx])
+            linear_projections.append(linear_proj)
```
```suggestion
            linear_projections.append(SegformerMLP(config, input_dim=config.hidden_sizes[stage_idx]))
```

vasqu commented on src/transformers/models/segformer/modular_segformer.py:
```diff
@@ -247,4 +260,556 @@ def _preprocess(
+        for stage_idx in range(config.num_encoder_blocks):
+            linear_proj = SegformerMLP(config, input_dim=config.hidden_sizes[stage_idx])
+            linear_projections.append(linear_proj)
+        self.linear_c = nn.ModuleList(linear_projections)
```
linear_c --> would be nice to rename

vasqu commented on src/transformers/models/segformer/modular_segformer.py:
```diff
@@ -247,4 +260,556 @@ def _preprocess(
+        if labels is not None and self.config.num_labels < 1:
+            raise ValueError(f"Number of labels should be >=0: {self.config.num_labels}")
+
+        # Track whether the user requested hidden states before we override the kwarg.
+        output_hidden_states = kwargs.get("output_hidden_states", getattr(self.config, "output_hidden_states", False))
```
Can we use the filter decorator instead? Or at least add a TODO here as well

vasqu commented on src/transformers/models/swin/modular_swin.py:
```diff
@@ -0,0 +1,1119 @@
+        value_states = self.v_proj(hidden_states).view(hidden_shape).transpose(1, 2)
+
+        # Combine relative position bias with the cyclic-shift attention mask for SW-MSA
+        relative_position_bias = self.relative_position_bias()  # 1, num_heads, ws*ws, ws*ws
```
Kind of reopening re https://github.com/huggingface/transformers/pull/41693/changes#r3063423265

Imo, we should buffer this instead of calling the forward each time - this won't change within the module

yonigozlan replied:
`relative_position_bias_table` is a learnable parameter, so if we want to support training for this model, I don't think we can buffer more than the index table no?

vasqu commented on src/transformers/models/swin/modular_swin.py:
```diff
@@ -0,0 +1,1119 @@
+    }
+
+    @torch.no_grad()
+    def _init_weights(self, module):
```
Missing super call or via PreTrained model? Should help avoid a few calls

vasqu commented on src/transformers/models/swin/modular_swin.py:
```diff
@@ -0,0 +1,1119 @@
+        super().__init__(config)
+
+        self.num_features = [config.embed_dim] + [int(config.embed_dim * 2**i) for i in range(len(config.depths))]
+        self.swin = SwinModel(config, add_pooling_layer=False)
```
Yea we really should double check re base model prefix if this is correct for some models or not

yonigozlan replied:
since we have base_model_prefix="swin", this should be good?

vasqu commented on src/transformers/models/vit_mae/modular_vit_mae.py:
```diff
@@ -0,0 +1,778 @@
+logger = logging.get_logger(__name__)
+
+
+def build_2d_sinusoidal_position_embedding(
```
Big difference is the cls token? Or could we just import it from a different model?

yonigozlan replied:
i think this is the first (chronologically) model in the library that has these embeddings, and all other models inherit from it

vasqu commented on src/transformers/models/vit_mae/modular_vit_mae.py:
```diff
@@ -0,0 +1,778 @@
+        self.patch_size = config.patch_size
+        self.config = config
+
+    def initialize_weights(self):
```
Ok this is a bit weird, but it is also so complicated that we kind of can't avoid it without clustering the init weights function itself

vasqu commented on src/transformers/models/vit_mae/modular_vit_mae.py:
```diff
@@ -0,0 +1,778 @@
+        # timm's trunc_normal_(std=.02) is effectively normal_(std=0.02) as cutoff is too big (2.)
+        init.normal_(self.cls_token, std=self.config.initializer_range)
+
+    def interpolate_pos_encoding(self, embeddings: torch.Tensor, height: int, width: int) -> torch.Tensor:
```
Might be wrong but could we inherit this from a different model?

vasqu commented on src/transformers/models/vit_mae/modular_vit_mae.py:
```diff
@@ -0,0 +1,778 @@
+
+        return torch.cat((class_pos_embed, patch_pos_embed), dim=1)
+
+    def random_masking(self, sequence, noise=None):
```
typing

vasqu commented on src/transformers/models/vit_mae/modular_vit_mae.py:
```diff
@@ -0,0 +1,778 @@
+        # masking: length -> length * config.mask_ratio
+        embeddings, mask, ids_restore = self.random_masking(embeddings, noise)
+
+        # append cls token
```
```suggestion
        # prepend cls token
```
nit

vasqu commented on src/transformers/models/vit_mae/modular_vit_mae.py:
```diff
@@ -0,0 +1,778 @@
+        self.config = decoder_config
+        self.initialize_weights(num_patches)
+
+    def interpolate_pos_encoding(self, embeddings: torch.Tensor) -> torch.Tensor:
```
Same re seems familiar to inherit maybe?

yonigozlan replied:
This decoder is quite specific to vit mae, so I don't think we can inherit cleanly from anything here :/

vasqu commented on src/transformers/models/vit_mae/modular_vit_mae.py:
```diff
@@ -0,0 +1,778 @@
+        interpolate_pos_encoding: bool = False,
+    ) -> ViTMAEDecoderOutput:
+        # Embed tokens
+        x = self.decoder_embed(hidden_states)
+
+        # Append mask tokens to sequence
+        mask_tokens = self.mask_token.repeat(x.shape[0], ids_restore.shape[1] + 1 - x.shape[1], 1)
+        x_ = torch.cat([x[:, 1:, :], mask_tokens], dim=1)
+
+        # Unshuffle
+        x_ = torch.gather(
+            x_,
+            dim=1,
+            index=ids_restore.unsqueeze(-1).repeat(1, 1, x.shape[2]).to(x_.device),
+        )
+        x = torch.cat([x[:, :1, :], x_], dim=1)
```
We should avoid the x vars, and rather use hidden states / residuals etc

vasqu commented on src/transformers/models/vit_mae/modular_vit_mae.py:
```diff
@@ -0,0 +1,778 @@
+
+@auto_docstring
+class ViTMAEPreTrainedModel(ViTPreTrainedModel):
+    config: ViTMAEConfig
+    base_model_prefix = "vit"
+    main_input_name = "pixel_values"
+    input_modalities = ("image",)
+    supports_gradient_checkpointing = True
+    _no_split_modules = ["ViTMAEEmbeddings", "ViTMAELayer", "ViTMAEDecoder"]
+    _supports_sdpa = True
+    _supports_flash_attn = True
+    _supports_flex_attn = True
+    _supports_attention_backend = True
+    _can_compile_fullgraph = True
+    _can_record_outputs = {
+        "hidden_states": ViTMAELayer,
+        "attentions": ViTMAEAttention,
+    }
```
Doubt we need to overwrite everything

vasqu commented on src/transformers/models/vit_mae/modular_vit_mae.py:
```diff
@@ -0,0 +1,778 @@
+    }
+
+    @torch.no_grad()
+    def _init_weights(self, module):
```
Similar super call might reduce some stuff here

vasqu commented on src/transformers/models/vit_mae/modular_vit_mae.py:
```diff
@@ -0,0 +1,778 @@
+
+@auto_docstring
+class ViTMAEModel(ViTMAEPreTrainedModel):
+    def __init__(self, config: ViTMAEConfig):
```
Init looks inheritable nit

vasqu commented on src/transformers/models/vit_mae/modular_vit_mae.py:
```diff
@@ -0,0 +1,778 @@
+        self.post_init()
+
+    def get_input_embeddings(self):
+        return self.vit.embeddings.patch_embeddings
```
self.vit.get_input_embeddings()?

yonigozlan replied:
I think I can remove it fully

vasqu commented on src/transformers/models/vit_mae/modular_vit_mae.py:
```diff
@@ -0,0 +1,778 @@
+            num_patches_w,
+            patch_size,
+        )
+        patchified_pixel_values = torch.einsum("nchpwq->nhwpqc", patchified_pixel_values)
```
Isn't that a permute? Would avoid the einsum tbh

vasqu commented on src/transformers/models/vit_mae/modular_vit_mae.py:
```diff
@@ -0,0 +1,778 @@
+            patch_size,
+            num_channels,
+        )
+        patchified_pixel_values = torch.einsum("nhwpqc->nchpwq", patchified_pixel_values)
```
same

vasqu commented on src/transformers/models/vit_mae/modular_vit_mae.py:
```diff
@@ -0,0 +1,778 @@
+        return pixel_values
+
+    def forward_loss(
+        self,
+        pixel_values,
+        pred,
+        mask,
+        interpolate_pos_encoding: bool = False,
```
typing

vasqu commented on src/transformers/models/vit_mae/modular_vit_mae.py:
```diff
@@ -0,0 +1,778 @@
+        )
+        logits = decoder_outputs.logits
+
+        loss = self.forward_loss(pixel_values, logits, mask, interpolate_pos_encoding=interpolate_pos_encoding)
```
Not a fan of this tbh - do we need the separate function?

vasqu commented on src/transformers/models/vit_msn/modular_vit_msn.py:
```diff
@@ -0,0 +1,226 @@
+    base_model_prefix = "vit"
+
+    @torch.no_grad()
+    def _init_weights(self, module):
```
same re super maybe reducing it a bit

vasqu commented on src/transformers/models/vit_msn/modular_vit_msn.py:
```diff
@@ -0,0 +1,226 @@
+
+@auto_docstring
+class ViTMSNModel(ViTMSNPreTrainedModel):
+    def __init__(self, config: ViTMSNConfig, use_mask_token: bool = False) -> None:
```
Could still inherit with modular imo, at least for the init

vasqu commented on src/transformers/models/vivit/modular_vivit.py:
```diff
@@ -0,0 +1,447 @@
+
+    def forward(self, pixel_values: torch.Tensor) -> torch.Tensor:
+        # permute to (batch_size, num_channels, num_frames, height, width) for Conv3d
+        pixel_values = pixel_values.permute(0, 2, 1, 3, 4)
```
My beloved transpose

vasqu commented on src/transformers/models/vivit/modular_vivit.py:
```diff
@@ -0,0 +1,447 @@
+    """
+
+    def __init__(self, config: VivitConfig):
+        nn.Module.__init__(self)
```
can use super if we dont inherit

vasqu commented on src/transformers/models/vivit/modular_vivit.py:
```diff
@@ -0,0 +1,447 @@
+        self.patch_size = config.tubelet_size[1:]
+        self.image_size = (config.image_size, config.image_size)
+
+    def interpolate_pos_encoding(self, embeddings: torch.Tensor, height: int, width: int) -> torch.Tensor:
```
Yea idk, would be cool to double check again but you definitely have a better overview on those

vasqu commented on src/transformers/models/vivit/modular_vivit.py:
```diff
@@ -0,0 +1,447 @@
+    input_modalities = ("video",)
+    supports_gradient_checkpointing = True
+    _no_split_modules = ["VivitEmbeddings", "VivitLayer"]
+    _supports_sdpa = True
+    _supports_flash_attn = True
+    _supports_flex_attn = True
+    _supports_attention_backend = True
+    _can_compile_fullgraph = True
```
Flags can be avoided a lot prolly

vasqu commented on src/transformers/models/vivit/modular_vivit.py:
```diff
@@ -0,0 +1,447 @@
+
+    @torch.no_grad()
+    def _init_weights(self, module):
+        """Initialize the weights"""
```
super

vasqu commented on src/transformers/models/vivit/modular_vivit.py:
```diff
@@ -0,0 +1,447 @@
+
+@auto_docstring
+class VivitModel(VivitPreTrainedModel):
+    def __init__(self, config: VivitConfig, add_pooling_layer: bool = True):
```
init inheritance

vasqu commented on tests/models/swin/test_modeling_swin.py:
```diff
@@ -223,7 +227,7 @@ class SwinModelTest(ModelTesterMixin, PipelineTesterMixin, unittest.TestCase):
         (
             SwinModel,
-            SwinBackbone,
+            # SwinBackbone,
```
Should be readded?

vasqu reviewed (approved):
Careful approval because the core is there. We have a few small exceptions here and there but I think this comes with the sheer size of models to handle and their legacy behavior.

Let's carefully go through those comments, make a mega run-slow and then it should be ready to merge

vasqu reviewed (approved):
Careful approval because the core is there. We have a few small exceptions here and there but I think this comes with the sheer size of models to handle and their legacy behavior.

Let's carefully go through those comments, make a mega run-slow and then it should be ready to merge

yonigozlan committed:
Address review

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into refactor-vit

yonigozlan committed:
fix repo

yonigozlan commented:
Addressed all the comments @vasqu !

Some small tests are failing which should be easy to fix (doing that tomorrow), but the rest should be good!
Do you want to have another look before I run all the slow tests?

vasqu commented on src/transformers/models/lw_detr/modular_lw_detr.py:
```diff
@@ -225,11 +222,21 @@ def validate_architecture(self):
         self.k_proj = nn.Linear(config.hidden_size, config.num_attention_heads * self.head_dim, bias=False)
+        self.v_proj = nn.Linear(config.hidden_size, config.num_attention_heads * self.head_dim, bias=config.qkv_bias)
+        self.o_proj = nn.Linear(config.num_attention_heads * self.head_dim, config.hidden_size, bias=True)
         self.num_key_value_groups = 1
```
Hmm, why do we still not use the super and only override the `self.attention_dropout = config.dropout_prob`

yonigozlan replied:
Forgot this one, thanks!

vasqu reviewed (approved):
I went over the changes and only found 1 small thing; trusting you on the rest. I guess you should sync with main as well then run the slow tests

yonigozlan committed:
Fix after review

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into refactor-vit

yonigozlan committed:
fix modular

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into refactor-vit

yonigozlan commented:
run-slow: aimv2, audio_spectrogram_transformer, beit, bit, clap, conditional_detr, convnext, convnextv2, cvt, d_fine, dab_detr, data2vec, deit, dinat, dinov2, dinov2_with_registers, dinov3_convnext, dinov3_vit, donut, dpt, eomt, eomt_dinov3, flava, florence2, focalnet, glpn, grounding_dino, hiera, ijepa, lw_detr, maskformer, mgp_str, mm_grounding_dino, pixio, poolformer, pp_doclayout_v2, pp_doclayout_v3, pp_lcnet, pp_lcnet_v3, pp_ocrv5_mobile_rec, pp_ocrv5_server_rec, pvt, pvt_v2, qianfan_ocr, resnet, rt_detr, rt_detr_v2, segformer, seggpt, slanet, swiftformer, swin, swin2sr, swinv2, timesformer, uvdoc, videomae, videomt, vilt, vit, vit_mae, vit_msn, vitdet, vitpose_backbone, vivit, vjepa2, x_clip, yolos

yonigozlan committed:
fix depth pro

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into refactor-vit

yonigozlan committed:
Fix conversion mappings

yonigozlan committed:
fix style

yonigozlan committed:
Merge branch 'main' into refactor-vit

yonigozlan committed:
Fix failing tests

yonigozlan committed:
Merge branch 'refactor-vit' of https://github.com/yonigozlan/transformers into refactor-vit

yonigozlan commented:
run-slow: aimv2, audio_spectrogram_transformer, beit, bit, clap, conditional_detr, convnext, convnextv2, cvt, d_fine, dab_detr, data2vec, deit, dinat, dinov2, dinov2_with_registers, dinov3_convnext, dinov3_vit, donut, dpt, eomt, eomt_dinov3, flava, florence2, focalnet, glpn, grounding_dino, hiera, ijepa, lw_detr, maskformer, mgp_str, mm_grounding_dino, pixio, poolformer, pp_doclayout_v2, pp_doclayout_v3, pp_lcnet, pp_lcnet_v3, pp_ocrv5_mobile_rec, pp_ocrv5_server_rec, pvt, pvt_v2, qianfan_ocr, resnet, rt_detr, rt_detr_v2, segformer, seggpt, slanet, swiftformer, swin, swin2sr, swinv2, timesformer, uvdoc, videomae, videomt, vilt, vit, vit_mae, vit_msn, vitdet, vitpose_backbone, vivit, vjepa2, x_clip, yolos

yonigozlan committed:
fix ci

yonigozlan committed:
Fix two slow failing tests
